### PR TITLE
[Refactor] Optimize the jdbc_scanner.cpp _precheck_data_type method

### LINKED ISSUE #47725: Optimize the jdbc_scanner.cpp _precheck_data_type method, which contains too many IFs

### PR DESCRIPTION
**Pull request by XinzhuangL**
## What I'm doing:

Fixes #47725 

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [x] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/exec/jdbc_scanner.cpp [hunk 1]
```diff
@@ -139,6 +140,7 @@ Status JDBCScanner::_init_jdbc_scan_context(RuntimeState* state) {
     int connection_timeout_ms =
             state->query_options().__isset.query_timeout ? state->query_options().query_timeout * 1000 : 30 * 1000;
 
+    LOG(INFO) << "Here is before NewObject";
```
Comment (XinzhuangL) [inline be/src/exec/jdbc_scanner.cpp:143]: Debug info. delete later

File: be/src/types/type_checker_manager.cpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    registerChecker("[B", std::make_unique<ByteArrayTypeChecker>());
+}
+
+TypeCheckerManager& TypeCheckerManager::getInstance() {
+    static TypeCheckerManager instance;
+    return instance;
+}
+
+void TypeCheckerManager::registerChecker(const std::string& java_class, std::unique_ptr<TypeChecker> checker) {
+    _checkers[java_class] = std::move(checker);
```
Comment (kevincai) [inline be/src/types/type_checker_manager.cpp:55]:
```suggestion
    _checkers.emplace(java_class, std::move(checker));
```

File: be/src/types/checker/type_checker.h [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+class ByteArrayTypeChecker : public TypeChecker {
+    StatusOr<LogicalType> check(const std::string& java_class, const SlotDescriptor* slot_desc) const override;
+};
+
+class DefaultTypeChecker : public TypeChecker {
+    StatusOr<LogicalType> check(const std::string& java_class, const SlotDescriptor* slot_desc) const override;
+};
+
+} // namespace starrocks
\ No newline at end of file
```
[the remark this answers is not in the document]
Comment (kevincai) [reply be/src/types/checker/type_checker.h:122]: valid comments.

File: be/src/types/checker/type_checker.cpp [hunk 4]
```diff
@@ -0,0 +1,319 @@
...
+
+#include <fmt/format.h>
+
+namespace starrocks {
+// java.lang.Byte
+
+StatusOr<LogicalType> ByteTypeChecker::check(const std::string& java_class,
+                                             const SlotDescriptor* slot_desc) const {
+    auto type = slot_desc->type().type;
+    if (type != TYPE_BOOLEAN && type != TYPE_TINYINT && type != TYPE_SMALLINT && type != TYPE_INT &&
```
Comment (kevincai) [inline be/src/types/checker/type_checker.cpp:25]: turn to switch-case, which may give a better human readable code style?

File: be/src/types/checker/type_checker.cpp [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+                            slot_desc->col_name()));
+    }
+    if (type == TYPE_BOOLEAN) {
+        return TYPE_BOOLEAN;
+    }
+    return TYPE_TINYINT;
+}
+
+// com.clickhouse.data.value.UnsignedByte
+
```
Comment (kevincai) [inline be/src/types/checker/type_checker.cpp:39]: useless empty line.

**Review by kevincai — COMMENTED** [review]
It would be awesome to add extensive unit test to cover the type check if there is no such cases yet.

File: be/src/types/checker/type_checker.cpp (hunk 5 above)
Comment (XinzhuangL) [reply be/src/types/checker/type_checker.cpp:39]: I will clean up the code and comments later

File: be/src/types/type_checker_manager.cpp (hunk 2 above)
Comment (XinzhuangL) [reply be/src/types/type_checker_manager.cpp:55]: I'll add the unit tests later

File: be/src/types/checker/type_checker.h [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+namespace starrocks {
+class TypeChecker {
+public:
+    virtual ~TypeChecker() = default;
+    virtual StatusOr<LogicalType> check(const std::string& java_class, const SlotDescriptor* slot_desc) const = 0;
+};
+
+class ByteTypeChecker : public TypeChecker {
+public:
```
Comment (zombee0) [inline be/src/types/checker/type_checker.h:31]: just a suggestion, maybe macro definition can help make code more clear.

**Review by zombee0 — COMMENTED** [review]
plz add some ut.

**Review by kevincai — APPROVED** [review]

**Review by zombee0 — APPROVED** [review]
LGTM

**Comment by zombee0** [comment]
@stdpain ptal

**Comment by stdpain** [comment]
From a functional perspective, this patch can be accepted. However, we can consider maintaining this relationship through a configuration file (xml or other). The hard-coded method is still too heavy and not flexible enough.

**Review by stdpain — APPROVED** [review]

**Comment by XinzhuangL** [comment]
> From a functional perspective, this patch can be accepted. However, we can consider maintaining this relationship through a configuration file (xml or other). The hard-coded method is still too heavy and not flexible enough.

Indeed

**Pull request merged**